### DISCUSSION TITLE
Add Button Links

### DISCUSSION THREAD

**Discussion by thelicato** [outline/outline#2940] [Ideas]
Since with Outline we have a Knowledge Base with a **tree-structure** with nested documents I think it would be better (from a UX perspective) to have the ability to add a **Button Link** in a document to link to other documents. I would use it to link to Nested Documents, but everyone could use it for different purposes. A single link is **barely visible** and the list of Nested Documents at the bottom of the page does not give me the structuring I want. 
I want to also provide you an example of what I am suggesting. Take for example a *GitBook* Knowledge Base, let's say [HackTricks](https://book.hacktricks.xyz/); in particular let's pick a random page that satisfies what I am saying, here is the result:

<img width="850" alt="image" src="https://user-images.githubusercontent.com/20096272/149486894-972d5d5f-8b0f-400b-884f-2d7a98c47294.png">

Even though on the left side I have the full structure as in Outline in the "**main**" document I see a beautiful button in a place that has been decided by the author of the document that gives me the ability to quickly go to a Nested Document. 

Let me know what you think and, as usual, **kudos** for all the work you do!

**Comment by tommoor** [comment]
It's a good idea, now we've integrated the editor into the codebase it opens up possibilities like this – previously everything in the editor had to be serializable to Markdown.

**Reply by thelicato** [reply]
Since there are more important things to implement I understand that this is not your top priority. If you give me some hints about how it should be implemented I can work on it and make a PR.

**Reply by thelicato** [reply]
Hi, maybe you lost my previous comment. I really want to help you to add this.

**Reply by tommoor** [reply]
Some of the work on https://github.com/outline/outline/pull/3031 will help with this as it's a very similar mechanic – file attachments are "special" links, the styling could also be very similar. 

You could take a look through that branch to get an understanding of the changes in the editor and perhaps base your work from there.

**Reply by thelicato** [reply]
Hi, It's been a while but I finally found the time to add this feature. I'm just having some trouble because I'd like to implement it the same way Gitbook does:
1. The user selects the "Page Link" component
2. The user searches for another document in the KB
3. The user clicks on the page he wants to link
4. The page link is created

Here are some screenshots:
![Screenshot from 2022-10-31 09-59-31](https://user-images.githubusercontent.com/20096272/198970873-6f6c3c53-9aa5-4bfa-81b0-a51c6f42b4e2.png)
![Screenshot from 2022-10-31 09-59-39](https://user-images.githubusercontent.com/20096272/198970890-3571b1eb-d797-4f69-b0d9-95a7685ebf1b.png)

So in Outline I would like to have something like the following:
![Screenshot from 2022-10-31 10-03-12](https://user-images.githubusercontent.com/20096272/198971206-a23c5f5d-e07a-4dbf-a7b5-9833dc4c35fb.png)
![Screenshot from 2022-10-31 10-03-33](https://user-images.githubusercontent.com/20096272/198971220-866ebffc-be00-42af-b509-6c54dd48533e.png)

In conclusion: I would like to have the same popup that appears when a **normal link** is added (but that indexes just the inner pages of Outline) and then after another document is selected the page link is created. I'm having some trouble with the code since the codebase is very big and seems that a lot of things are decoupled. Can you help me?

**Reply by tommoor** [reply]
Feels dependent on https://github.com/outline/outline/issues/4273 and for the menu and https://github.com/outline/outline/issues/3000 for storing the node type as currently everything in the editor has to be losslessly serializable to Markdown until that's completed.

**Reply by thelicato** [reply]
Hi, I just did it! I opened a PR. I've added a new component called "PageLinkToolbar" that handles the creation of this node. The link to the PR is this: https://github.com/outline/outline/pull/4367

I've also added a few optional props to the **LinkEditor** to avoid creating a new component too similar to it. As in GitBook the user can only link other documents in the KB (not external links).

**Reply by thelicato** [reply]
Hi, I know you are probably busy with a lot of stuff right now. I deployed an instance of my fork for personal use. The new component seems to work pretty well, the only problem is that it automatically gives to the **Page Link** the name of the linked document and therefore:
- If the document name changes the component keeps the old name
- If the document gets deleted the Page Link remains

In GitBook, on the other hand, when a document is updated or deleted the changes are automatically reflected into the Page Link. I don't know if it is currently possible to implement such behaviour in Outline. Nevertheless I think that right now the component is still usable with these issues
